src/lib/chawathe98_matcher.rs

```diff
@@ -46,35 +46,2262 @@
 
 use std::fmt::Debug;
 
-use ast::Arena;
-use matchers::{MappingStore, MatchTrees};
+use ast::{Arena, FromNodeId, NodeId, ToNodeId};
 
-/// The Chawathe98Config matcher needs no configuration.
+use std::cell::RefCell;
+use std::collections::HashSet;
+use std::cmp;
+
+/// Not needed by matching algorithms, but useful for debugging.
+#[derive(Debug, Clone, Eq, Hash, PartialEq)]
+pub enum EdgeType {
+    /// Check the type of edge update which is update
+    UPDATE,
+    /// Check the type of edge update which is insert
+    INSERT,
+    /// Check the type of edge update which is delete
+    DELETE,
+    /// Check the type of edge update which is move
+    MOVE,
+    /// Check the type of edge update which is copy
+    COPY,
+    /// Check the type of edge update which is glue
+    GLUE,
+    /// Check the type of edge update which is null
+    NULL,
+    /// Check the type of edge update which is Ok
+    OK,
+}
+#[derive(Debug, Clone)]
+/// The intial cost of each edge
+pub struct CostEdge {
+    /// Check the cost of edge update which is update
+    pub cost_update: usize,
```

Comment:
Because the struct here is called `CostEdge` we know it contains the costs of edges, so I think you can call the struct members `update` rather than `cost_update`, etc.

Reply:
Also, I think a `usize` maybe isn't quite what you want here, can a cost really be an arbitrary size? Probably a `u8` or `u16` would be OK?

Reply:
It's worth thinking carefully about the right size, and using a bigger-rather-than-smaller size if you're not sure. I agree that `usize` is unlikely to be the right choice here, though.